src/System.Text.RegularExpressions/ref/System.Text.RegularExpressions.cs

```diff
@@ -87,7 +87,7 @@ public partial class GroupCollection : System.Collections.Generic.ICollection<Sy
         void System.Collections.IList.Insert(int index, object value) { }
         void System.Collections.IList.Remove(object value) { }
         void System.Collections.IList.RemoveAt(int index) { }
-        public bool TryGetValue(string key, out System.Text.RegularExpressions.Group value) { throw null; }
+        public bool TryGetValue(string key, [System.Diagnostics.CodeAnalysis.MaybeNullWhenAttribute(false)] out System.Text.RegularExpressions.Group value) { throw null; }
```

Comment:
`value` will never be null if this returns true, right?  If so, this should be `NotNullWhen(true)` rather than `MaybeNullWhen(false)`.

Reply:
`value` should also be `Group?` rather than `Group`.

Reply:
This is implementing [`IReadOnlyDictionary<_,_>.TryGetValue` which uses `MaybeNullWhen(false)`](https://github.com/dotnet/corefx/blob/ffb94ad2c2d7c14fd9be82d365a28a60725b4f9f/src/Common/src/CoreLib/System/Collections/Generic/IReadOnlyDictionary.cs#L13).

Reply:
`IReadOnlyDictionary<,>.TryGetValue` is generic, that is why we use `MaybeNullWhenFalse` rather than `NotNull` because we can't make a generic parameter nullable, and for the compiler generics are not nullable on the output, so therefore, we need to tell the compiler, value might be null if this is false. 

However, in this case since `Group` is a reference type, we should make it nullable and then since for the compiler it can be null, we need to add an attribute saying "it will not be null when we return false", hence we need to use `NotNullWhen(true)` because we know it will never be null if it returns true and we can assume it will be null otherwise.

Reply:
Yeah, I think we should use `[NotNullWhen(true)] Group? group`, and if the compiler complains, suppress the warning with a pragma.  Might be worth a separate discussion, though.

Reply:
I'm not sure I understand what the ramification of such a change would be. Does it mean that I'd get different annotations depending on whether I have upcasted to `IReadOnlyDictionary<_,_>`? Is that even desirable?

Reply:
Whether calling through the interface or the concrete type, the caller will need to use `out Group?` rather than `out Group`, or else they'll get a warning "Converting null literal or possible null value to non-nullable type".  This way, though, that's more obvious when looking at the method's signature, whether in docs or in IntelliSense.